File: src/implementation/ios/runtime/XTRViewController.m
//
//  XTRViewController.m
//  XTSample
//
//  Created by 崔明辉 on 2017/8/25.
//  Copyright © 2017年 UED Center, YY Inc. All rights reserved.
//

#import "XTRViewController.h"
#import "XTRUtils.h"

@interface XTRViewController ()

@property (nonatomic, strong) JSContext *context;
@property (nonatomic, strong) JSManagedValue *scriptObject;

@end

@implementation XTRViewController

+ (NSString *)name {
    return @"XTRViewController";
}

+ (XTRViewController *)create:(JSValue *)scriptObject {
    XTRViewController *viewController = [XTRViewController new];
    viewController.objectUUID = [[NSUUID UUID] UUIDString];
    viewController.context = scriptObject.context;
    viewController.scriptObject = [JSManagedValue managedValueWithValue:scriptObject andOwner:viewController];
    return viewController;
}

- (JSValue *)xtr_view {
    return [JSValue fromObject:self.view context:self.context];
}

- (void)xtr_setView:(JSValue *)view {
    self.view = [view toView];
}

- (void)loadView {
    if (self.scriptObject != nil) {
        JSValue *value = self.scriptObject.value;
        if (value != nil) {
            [value invokeMethod:@"loadView" withArguments:@[]];
        }
    }
}

- (void)viewDidLoad {
    [super viewDidLoad];
    self.automaticallyAdjustsScrollViewInsets = NO;
    if (self.scriptObject != nil) {
        JSValue *value = self.scriptObject.value;
        if (value != nil) {
            [value invokeMethod:@"viewDidLoad" withArguments:@[]];
        }
    }
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    if (self.scriptObject != nil) {
        JSValue *value = self.scriptObject.value;
        if (value != nil) {
            [value invokeMethod:@"viewWillAppear" withArguments:@[]];
        }
    }
}

- (void)viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];
    if (self.scriptObject != nil) {
        JSValue *value = self.scriptObject.value;
        if (value != nil) {
            [value invokeMethod:@"viewDidAppear" withArguments:@[]];
        }
    }
}

- (void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];
    if (self.scriptObject != nil) {
        JSValue *value = self.scriptObject.value;
        if (value != nil) {
            [value invokeMethod:@"viewWillDisappear" withArguments:@[]];
        }
    }
}

- (void)viewDidDisappear:(BOOL)animated {
    [super viewDidDisappear:animated];
    if (self.scriptObject != nil) {
        JSValue *value = self.scriptObject.value;
        if (value != nil) {
            [value invokeMethod:@"viewDidDisappear" withArguments:@[]];
        }
    }
}

- (void)viewWillLayoutSubviews {
    [super viewWillLayoutSubviews];
    if (self.scriptObject != nil) {
        JSValue *value = self.scriptObject.value;
        if (value != nil) {
            [value invokeMethod:@"viewWillLayoutSubviews" withArguments:@[]];
        }
    }
}

- (void)viewDidLayoutSubviews {
    [super viewDidLayoutSubviews];
    if (self.scriptObject != nil) {
        JSValue *value = self.scriptObject.value;
        if (value != nil) {
            [value invokeMethod:@"viewDidLayoutSubviews" withArguments:@[]];
        }
    }
}

- (JSValue *)xtr_parentViewController {
    return [JSValue fromObject:self.parentViewController context:self.context];
}

- (NSArray<JSValue *> *)xtr_childViewControllers {
    NSMutableArray *childViewControllers = [NSMutableArray array];
    for (UIViewController *viewController in self.childViewControllers) {
        [childViewControllers addObject:[JSValue fromObject:viewController context:self.context] ?: [NSNull null]];
    }
    return [childViewControllers copy];
}

- (void)xtr_addChildViewController:(JSValue *)childController {
    UIViewController *viewController = [childController toViewController];
    if (viewController) {
        [self addChildViewController:viewController];
    }
}

- (void)xtr_removeFromParentViewController {
    [self removeFromParentViewController];
}

- (void)willMoveToParentViewController:(UIViewController *)parent {
    [super willMoveToParentViewController:parent];
    if (self.scriptObject != nil) {
        JSValue *value = self.scriptObject.value;
        if (value != nil) {
            [value invokeMethod:@"willMoveToParentViewController" withArguments:parent != nil ? @[
                                                                                                  [JSValue fromObject:parent
                                                                                                              context:self.context]
                                                                                                  ] : @[]];
        }
    }
}

- (void)didMoveToParentViewController:(UIViewController *)parent {
    [super didMoveToParentViewController:parent];
    if (self.scriptObject != nil) {
        JSValue *value = self.scriptObject.value;
        if (value != nil) {
            [value invokeMethod:@"didMoveToParentViewController" withArguments:parent != nil ? @[
                                                                                                  [JSValue fromObject:parent
                                                                                                              context:self.context]
                                                                                                  ] : @[]];
        }
    }
}

- (JSValue *)xtr_navigationController {
    return [JSValue fromObject:self.navigationController context:self.context];
}

@end

File: samples/Android/xtruntime/src/main/java/com/opensource/xtruntime/XTRViewController.kt
package com.opensource.xtruntime

import com.eclipsesource.v8.V8
import com.eclipsesource.v8.V8Array
import com.eclipsesource.v8.V8Object
import java.util.*

/**
 * Created by cuiminghui on 2017/9/5.
 */
open class XTRViewController: XTRComponent() {

    override val name: String = "XTRViewController"

    override fun v8Object(): V8Object? {
        XTRImage.runtime = xtrContext.v8Runtime
        val v8Object = V8Object(xtrContext.v8Runtime)
        v8Object.registerJavaMethod(this, "createScriptObject", "createScriptObject", arrayOf(V8Object::class.java))
        return v8Object
    }

    fun createScriptObject(scriptObject: V8Object): V8Object {
        return InnerObject(xtrContext.autoRelease(scriptObject.twin()), xtrContext).requestV8Object(xtrContext.v8Runtime)
    }

    open class InnerObject(override var scriptObject: V8Object?, protected val xtrContext: XTRContext): XTRObject {

        override val objectUUID: String = UUID.randomUUID().toString()

        override fun requestV8Object(runtime: V8): V8Object {
            val v8Object = super.requestV8Object(runtime)
            v8Object.registerJavaMethod(this, "xtr_view", "xtr_view", arrayOf())
            v8Object.registerJavaMethod(this, "xtr_setView", "xtr_setView", arrayOf(V8Object::class.java))
            v8Object.registerJavaMethod(this, "xtr_parentViewController", "xtr_parentViewController", arrayOf())
            v8Object.registerJavaMethod(this, "xtr_childViewControllers", "xtr_childViewControllers", arrayOf())
            v8Object.registerJavaMethod(this, "xtr_addChildViewController", "xtr_addChildViewController", arrayOf(V8Object::class.java))
            v8Object.registerJavaMethod(this, "xtr_removeFromParentViewController", "xtr_removeFromParentViewController", arrayOf())
            v8Object.registerJavaMethod(this, "xtr_navigationController", "xtr_navigationController", arrayOf())
            return v8Object
        }

        var view: XTRView.InnerObject? = null
            set(value) {
                field?.let {
                    it.viewDelegate = null
                }
                field = value
                field?.let {
                    it.viewDelegate = this
                }
            }

        fun xtr_view(): Any? {
            view?.let {
                return XTRUtils.fromObject(xtrContext, it)
            }
            return V8.getUndefined()
        }

        fun xtr_setView(value: V8Object) {
            (XTRUtils.toView(value) as? XTRView.InnerObject)?.let {
                this.view = it
                viewDidLoad()
            }
        }

        var parentViewController: XTRViewController.InnerObject? = null
            protected set

        var childViewControllers: List<XTRViewController.InnerObject> = listOf()
            protected set

        fun xtr_parentViewController(): Any? {
            return XTRUtils.fromObject(xtrContext, parentViewController)
        }

        fun xtr_childViewControllers(): V8Array? {
            return XTRUtils.fromObject(xtrContext, childViewControllers) as? V8Array
        }

        fun xtr_addChildViewController(childController: V8Object) {
            XTRUtils.toViewController(childController)?.let { childController ->
                if (childController.parentViewController == null) {
                    childController.willMoveToParentViewController(this)
                    childViewControllers.toMutableList()?.let {
                        it.add(childController)
                        childController.parentViewController = this
                        childViewControllers = it.toList()
                    }
                    childController.didMoveToParentViewController(this)
                }
            }
        }

        fun xtr_addChildViewController(childController: XTRViewController.InnerObject) {
            if (childController.parentViewController == null) {
                childController.willMoveToParentViewController(this)
                childViewControllers.toMutableList()?.let {
                    it.add(childController)
                    childController.parentViewController = this
                    childViewControllers = it.toList()
                }
                childController.didMoveToParentViewController(this)
            }
        }

        fun xtr_removeFromParentViewController() {
            parentViewController?.let { parentViewController ->
                willMoveToParentViewController(null)
                parentViewController.childViewControllers.toMutableList()?.let {
                    it.remove(this)
                    parentViewController.childViewControllers = it.toList()
                }
                didMoveToParentViewController(null)
            }
            parentViewController = null
        }

        fun xtr_navigationController(): Any? {
            var current: XTRViewController.InnerObject? = this
            while (current != null) {
                (current as? XTRNavigationController.InnerObject)?.let {
                    return XTRUtils.fromObject(xtrContext, it)
                }
                current = current.parentViewController
            }
            return V8.getUndefined()
        }

        fun willMoveToParentViewController(parent: XTRViewController.InnerObject?) {
            parent?.let {
                xtrContext.invokeMethod(scriptObject, "willMoveToParentViewController", listOf(it))
                return
            }
            xtrContext.invokeMethod(scriptObject, "willMoveToParentViewController", null)
        }

        fun didMoveToParentViewController(parent: XTRViewController.InnerObject?) {
            parent?.let {
                xtrContext.invokeMethod(scriptObject, "didMoveToParentViewController", listOf(it))
                return
            }
            xtrContext.invokeMethod(scriptObject, "didMoveToParentViewController", null)
        }

        open fun viewDidLoad() {
            xtrContext.invokeMethod(scriptObject, "viewDidLoad", null)
        }

        open fun viewWillAppear() {
            xtrContext.invokeMethod(scriptObject, "viewWillAppear", null)
        }

        open fun viewDidAppear() {
            xtrContext.invokeMethod(scriptObject, "viewDidAppear", null)
        }

        open fun viewWillDisappear() {
            xtrContext.invokeMethod(scriptObject, "viewWillDisappear", null)
        }

        open fun viewDidDisappear() {
            xtrContext.invokeMethod(scriptObject, "viewDidDisappear", null)
        }

        open fun viewWillLayoutSubviews() {
            xtrContext.invokeMethod(scriptObject, "viewWillLayoutSubviews", null)
        }

        open fun viewDidLayoutSubviews() {
            xtrContext.invokeMethod(scriptObject, "viewDidLayoutSubviews", null)
        }

    }

}